service/test/BatchServerTest.cpp

```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
+        EXPECT_CALL(*m_pio_ptr, push_control(request.name, request.domain,
+                                             request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    EXPECT_CALL(*m_batch_status, receive_message())
+        .WillOnce(Return(BatchStatus::M_MESSAGE_TERMINATE))
+        .RetiresOnSaturation();
+
+    m_batch_server->run_batch();
+}
+
+/**
+ * @test Provides coverage for the case when not a valid message is read
+ *       by the server in BatchServerImp::event_loop()
+ */
+TEST_F(BatchServerTest, receive_message_default)
+{
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
+
+    for (const auto &request : m_signal_config) {
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
+        EXPECT_CALL(*m_pio_ptr, push_control(request.name, request.domain,
+                                             request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    EXPECT_CALL(*m_batch_status, receive_message())
+        .WillOnce(Return(127))  // a "random" number
+        .RetiresOnSaturation();
+
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_batch_server->run_batch(),
+        GEOPM_ERROR_RUNTIME,
+        "BatchServerImp::run_batch(): Received unknown response from client: "
+    );
+}
+
+/**
+ * @throws geopm Exception to simulate failure of read(2) system call in BatchStatusImp::receive_message()
+ */
+ACTION(batch_status_read_EINTR)
+{
+    throw geopm::Exception(
+        "BatchStatusImp: System call failed: " "read(2)",
+        EINTR,
+        __FILE__,
+        __LINE__
+    );
+}
+
+/**
+ * @test Testing BatchStatus::receive_message() to throw an exception inside of BatchServerImp::read_message()
+ *       This is needed to activate the catch blocks in BatchServerImp::read_message() and BatchServerImp::run_batch()
+ */
+TEST_F(BatchServerTest, receive_message_exception)
+{
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
+
+    for (const auto &request : m_signal_config) {
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```

Comment:
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.